Add EventHandler docs

sofroniewn opened the pull request:
# Description
This PR adds explanation docs for the EventHandler refactor #1376. It includes the amazing diagram from #1040 that @shanaxel42 made, and a link to this wikipedia article on the [model-view-presenter](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93presenter) which I thought might be helpful.

The doc is definitely aspirational and describes where we want to be, but I hope it will be helpful to people trying to understand our architecture. It will certainly need updating and refining as we go, but I think it would be nice to get in fairly soon after #1376 and then update it so we at least have something there.

Thoughts @jni @tlambert03 @ziyangczi @kevinyamauchi @shanaxel42? This is ready for review!

sofroniewn committed:
add docs

liu-ziyang commented on docs/source/explanations/event_handling.md:
````diff
@@ -0,0 +1,84 @@
+the ```EventHandler``` will receive this event and its payload and then will
+call any corresponding update methods on any object that has been registered
+as a listener. Update methods are linked to particular emitters via a naming
+convention ```_on_*change``` where the ```*``` corresponds to the name of the
````
```suggestion
convention ```_on_*_change``` where the ```*``` corresponds to the name of the
```

sofroniewn committed:
Update docs/source/explanations/event_handling.md

Co-authored-by: ziyangczi <41968256+ziyangczi@users.noreply.github.com>

willingc commented on docs/source/explanations/event_handling.md:
```diff
@@ -0,0 +1,84 @@
+napari is architected to have a strict separation of its model code and its 
+view code.
+
+- **models** directly manages data, logic, and rules of within napari,
```
```suggestion
- **models** directly manage data, logic, and rules of within napari,
```

willingc commented on docs/source/explanations/event_handling.md:
```diff
@@ -0,0 +1,84 @@
+- **views** contain a representation of information from a model and route
+user commands to an event handler via events.
+
+- **events** are emitted from both views and models in napari and all flow
```
```suggestion
- **events** are emitted from both views and models in napari and all events flow
```

willingc commented on docs/source/explanations/event_handling.md:
```diff
@@ -0,0 +1,84 @@
+are all being connected via an event handler. 
+
+This architecture is probably most similar to the [model-view-presenter](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93presenter) architecture. A unique aspect of napari
+that distinguishes it from other GUI applications is that advanced users expect 
```
```suggestion
that distinguishes it from other GUI applications is that advanced users expect: 
```

willingc commented on docs/source/explanations/event_handling.md:
```diff
@@ -0,0 +1,84 @@
+
+This architecture is probably most similar to the [model-view-presenter](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93presenter) architecture. A unique aspect of napari
+that distinguishes it from other GUI applications is that advanced users expect 
+to be able to interact directly with the napari model from interactive
```
```suggestion
- to be able to interact directly with the napari model from **interactive**
```

willingc commented on docs/source/explanations/event_handling.md:
````diff
@@ -0,0 +1,84 @@
+This architecture is probably most similar to the [model-view-presenter](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93presenter) architecture. A unique aspect of napari
+that distinguishes it from other GUI applications is that advanced users expect 
+to be able to interact directly with the napari model from interactive
+environments like ```IPython``` and ```Jupyter``` and to use napari models
````
```suggestion
environments like ```IPython``` and ```Jupyter``` 
- to use napari models
```

willingc commented on docs/source/explanations/event_handling.md:
````diff
@@ -0,0 +1,84 @@
+that distinguishes it from other GUI applications is that advanced users expect 
+to be able to interact directly with the napari model from interactive
+environments like ```IPython``` and ```Jupyter``` and to use napari models
+headless, without any views when performing batch analysis. This feature forces
````
```suggestion
headless, without any views when performing **batch** analysis.

These uses motivate
```

willingc commented on docs/source/explanations/event_handling.md:
```diff
@@ -0,0 +1,84 @@
+us to put more logic into our model files so that they can be used stand-alone.
+
+Right now napari just supports one front-end, built using [Qt](https://www.qt.io/),
+and one rendering library using [VisPy](http://vispy.org/), but in the future
```
```suggestion
and one rendering library using [VisPy](http://vispy.org/). In the future
```

willingc commented on docs/source/explanations/event_handling.md:
````diff
@@ -0,0 +1,84 @@
+connected to an ```EventHandler``` that contains various named ```EventEmitter```
+objects, each of which is capable of emitting events that can carry a single
+value as a payload. For example you can emit a custom colormap string from
+the ```colormap``` emitter as follows:
````
```suggestion
the ```colormap``` event emitter as follows:
```

willingc commented on docs/source/explanations/event_handling.md:
````diff
@@ -0,0 +1,84 @@
+layer.events.colormap('my_colormap')
+```
+
+If this ```EventEmitterGroup``` was connected to an ```EventHandler``` then
````
```suggestion
By connecting an ```EventEmitterGroup``` to an ```EventHandler```, events can trigger an event handler to perform actions. When the event is emitted,
```

willingc commented on docs/source/explanations/event_handling.md:
````diff
@@ -0,0 +1,84 @@
+```
+
+If this ```EventEmitterGroup``` was connected to an ```EventHandler``` then
+the ```EventHandler``` will receive this event and its payload and then will
````
```suggestion
the ```EventHandler``` will receive this event and its payload and act on the event by
```

willingc commented on docs/source/explanations/event_handling.md:
````diff
@@ -0,0 +1,84 @@
+
+If this ```EventEmitterGroup``` was connected to an ```EventHandler``` then
+the ```EventHandler``` will receive this event and its payload and then will
+call any corresponding update methods on any object that has been registered
````
```suggestion
calling any corresponding update methods on any object that has been registered
```

willingc commented on docs/source/explanations/event_handling.md:
````diff
@@ -0,0 +1,84 @@
+updating patterns. 
+
+As we expect napari users to be able to interact with our model object directly
+we use ```setters``` on our attributes to emit events and ```_on_*_change```
````
```suggestion
we use ```setters``` on model attributes to emit events and ```_on_*_change```
```

willingc commented on docs/source/explanations/event_handling.md:
````diff
@@ -0,0 +1,84 @@
+we use ```setters``` on our attributes to emit events and ```_on_*_change```
+methods to actually perform updates. When instantiating our models the
+necessary event handlers are created and attached to them so that they can
+be used without additional imports. The separation of our setters, event
````
```suggestion
be used without additional imports. The clear separation of our setters, event
```

willingc reviewed (approved):
@sofroniewn This looks great. I've made some suggestions to improve clarity. Event handling is something that new contributors and users find conceptually challenging. In my experience a bit more precision (re: pronoun references) aids understanding.

sofroniewn committed:
Apply suggestions from code review

These edits are fantastic @willingc!! So helpful to have someone go through the PR like this. We might not actually merge this as is, as we've been having some discussion about the final event architecture (see #1409), but I might start a new PR based off of this one, but with some of the details/ diagram changed!!

Also you're right the windows CI fail is independent of this PR. There's some attempt to fix it in #1404, but we need to work on that more

Co-authored-by: Carol Willing <carolcode@willingconsulting.com>
